xray/commands/audit/scan.go

```diff
@@ -0,0 +1,270 @@
+	spec          *spec.SpecFiles
+	threads       int
+	indexerPath   string
+}
+
+func (scanCmd *XrBinariesScanCommand) SetThreads(threads int) *XrBinariesScanCommand {
+	scanCmd.threads = threads
+	return scanCmd
+}
+
+func (scanCmd *XrBinariesScanCommand) SetServerDetails(server *config.ServerDetails) *XrBinariesScanCommand {
+	scanCmd.serverDetails = server
+	return scanCmd
+}
+
+func (scanCmd *XrBinariesScanCommand) SetSpec(spec *spec.SpecFiles) *XrBinariesScanCommand {
+	scanCmd.spec = spec
+	return scanCmd
+}
+
+func (scanCmd *XrBinariesScanCommand) ServerDetails() (*config.ServerDetails, error) {
+	return scanCmd.serverDetails, nil
+}
+
+func (scanCmd *XrBinariesScanCommand) IndexFile(filePath string) (*services.GraphNode, error) {
+	var indexerResults services.GraphNode
+	indexCmd := &coreutils.GeneralExecCmd{
+		ExecPath: scanCmd.indexerPath,
+		Command:  []string{IndexingCommand, filePath},
+	}
+	output, err := io.RunCmdOutput(indexCmd)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal([]byte(output), &indexerResults)
+	return &indexerResults, err
+
+}
+
+func (scanCmd *XrBinariesScanCommand) GetXrScanGraphResults(graph *services.GraphNode) (*services.ScanResponse, error) {
+	xrayManager, err := commands.CreateXrayServiceManager(scanCmd.serverDetails)
+	if err != nil {
+		return nil, err
+	}
+	params := services.NewXrayGraphScanParams()
+	params.Graph = graph
+	scanId, err := xrayManager.ScanGraph(params)
+	if err != nil {
+		return nil, err
+	}
+	scanResults, err := xrayManager.GetScanGraphResults(scanId)
+	if err != nil {
+		return nil, err
+	}
+	return scanResults, nil
+}
+
+func (scanCmd *XrBinariesScanCommand) Run() (err error) {
+	// First download Xray Indexer if needed
+	xrayManager, err := commands.CreateXrayServiceManager(scanCmd.serverDetails)
+	if err != nil {
+		return err
+	}
+	scanCmd.indexerPath, err = xrutils.DownloadIndexerIfNeeded(xrayManager)
+	if err != nil {
+		return err
+	}
+	resultsArr := make([][]*services.ScanResponse, scanCmd.threads)
+	fileProducerConsumer := parallel.NewRunner(scanCmd.threads, 20000, false)
+	fileProducerErrorsQueue := clientutils.NewErrorsQueue(1)
+	indexedFileProducerConsumer := parallel.NewRunner(scanCmd.threads, 20000, false)
+	indexedFileProducerErrorsQueue := clientutils.NewErrorsQueue(1)
+	// Start walk on the Filesystem "produce" files that match the given pattern
+	// while the consumer uses the indexer to index those files.
+	scanCmd.prepareScanTasks(fileProducerConsumer, indexedFileProducerConsumer, resultsArr, fileProducerErrorsQueue, indexedFileProducerErrorsQueue)
+	scanCmd.performScanTasks(fileProducerConsumer, indexedFileProducerConsumer, resultsArr)
+	err = fileProducerErrorsQueue.GetError()
+	if err != nil {
+		return err
+	}
+	err = indexedFileProducerErrorsQueue.GetError()
+	if err != nil {
+		return err
+	}
+	// Start "consume" the files that was indexed by sending graph scan request to Xray and
+	// waits for results.
+
+	// Handel saved scan results.
```

Comment:
Some code is still missing here, right?

Reply:
No, Just left-over comments I wrote before implementation.
Deleted.